// electron-builder-util/out/log.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.warn = warn;
exports.log = log;
exports.subTask = subTask;
exports.task = task;

var _chalk;

function _load_chalk() {
    return _chalk = require("chalk");
}

var _nodeEmoji;

function _load_nodeEmoji() {
    return _nodeEmoji = require("node-emoji");
}

class Logger {
    constructor(stream) {
        this.stream = stream;
        this.isTTY = process.stdout.isTTY;
    }
    warn(message) {
        if (this.isTTY) {
            this.log((0, (_nodeEmoji || _load_nodeEmoji()).get)("warning") + "  " + (0, (_chalk || _load_chalk()).yellow)(message));
        } else {
            this.log((0, (_chalk || _load_chalk()).yellow)(`Warning: ${message}`));
        }
    }
    log(message) {
        this.stream.write(`${message}\n`);
    }
    subTask(title, _promise) {
        return _promise;
    }
    task(title, _promise) {
        const promise = _promise;
        this.log(`${title}\n`);
        return promise;
    }
}
const logger = new Logger(process.stdout);
function warn(message) {
    logger.warn(message);
}
function log(message) {
    logger.log(message);
}
function subTask(title, promise) {
    return logger.subTask(title, promise);
}
function task(title, promise) {
    return logger.task(title, promise);
}
