// electron-builder/out/util/nodeHttpExecutor.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.NodeHttpExecutor = undefined;

var _bluebirdLstC;

function _load_bluebirdLstC() {
    return _bluebirdLstC = require("bluebird-lst-c");
}

var _bluebirdLstC2;

function _load_bluebirdLstC2() {
    return _bluebirdLstC2 = _interopRequireDefault(require("bluebird-lst-c"));
}

// only https proxy
let proxyFromNpm = (() => {
    var _ref = (0, (_bluebirdLstC || _load_bluebirdLstC()).coroutine)(function* () {
        let data = "";
        try {
            data = yield (0, (_fsExtraP || _load_fsExtraP()).readFile)(_path.join((0, (_os || _load_os()).homedir)(), ".npmrc"), "utf-8");
        } catch (ignored) {
            return null;
        }
        if (!data) {
            return null;
        }
        try {
            const config = (0, (_ini || _load_ini()).parse)(data);
            return config["https-proxy"] || config.proxy;
        } catch (e) {
            // used in nsis auto-updater, do not use .util.warn here
            console.warn(e);
            return null;
        }
    });

    return function proxyFromNpm() {
        return _ref.apply(this, arguments);
    };
})();
// only https url


let createAgent = (() => {
    var _ref2 = (0, (_bluebirdLstC || _load_bluebirdLstC()).coroutine)(function* () {
        let proxyString = process.env.npm_config_https_proxy || process.env.HTTPS_PROXY || process.env.https_proxy || process.env.npm_config_proxy;
        if (!proxyString) {
            proxyString = yield proxyFromNpm();
            if (!proxyString) {
                return null;
            }
        }
        const proxy = (0, (_url || _load_url()).parse)(proxyString);
        const proxyProtocol = proxy.protocol === "https:" ? "Https" : "Http";
        return require("tunnel-agent")[`httpsOver${proxyProtocol}`]({
            proxy: {
                port: proxy.port || (proxyProtocol === "Https" ? 443 : 80),
                host: proxy.hostname,
                proxyAuth: proxy.auth
            }
        });
    });

    return function createAgent() {
        return _ref2.apply(this, arguments);
    };
})();


var _https;

function _load_https() {
    return _https = _interopRequireWildcard(require("https"));
}

var _fsExtraP;

function _load_fsExtraP() {
    return _fsExtraP = require("fs-extra-p");
}

var _path = _interopRequireWildcard(require("path"));

var _os;

function _load_os() {
    return _os = require("os");
}

var _ini;

function _load_ini() {
    return _ini = require("ini");
}

var _electronBuilderHttp;

function _load_electronBuilderHttp() {
    return _electronBuilderHttp = require("electron-builder-http");
}

var _url;

function _load_url() {
    return _url = require("url");
}

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class NodeHttpExecutor extends (_electronBuilderHttp || _load_electronBuilderHttp()).HttpExecutor {
    download(url, destination, options) {
        var _this = this;

        return (0, (_bluebirdLstC || _load_bluebirdLstC()).coroutine)(function* () {
            if (options == null || !options.skipDirCreation) {
                yield (0, (_fsExtraP || _load_fsExtraP()).ensureDir)(_path.dirname(destination));
            }
            if (_this.httpsAgentPromise == null) {
                _this.httpsAgentPromise = createAgent();
            }
            const agent = yield _this.httpsAgentPromise;
            return yield new (_bluebirdLstC2 || _load_bluebirdLstC2()).default(function (resolve, reject) {
                _this.doDownload(url, destination, 0, options || {}, agent, function (error) {
                    if (error == null) {
                        resolve(destination);
                    } else {
                        reject(error);
                    }
                });
            });
        })();
    }
    addTimeOutHandler(request, callback) {
        request.on("socket", function (socket) {
            socket.setTimeout(60 * 1000, () => {
                callback(new Error("Request timed out"));
                request.abort();
            });
        });
    }
    doDownload(url, destination, redirectCount, options, agent, callback) {
        const parsedUrl = (0, (_url || _load_url()).parse)(url);
        // user-agent must be specified, otherwise some host can return 401 unauthorised
        const request = (_https || _load_https()).request({
            hostname: parsedUrl.hostname,
            path: parsedUrl.path,
            headers: {
                "User-Agent": "electron-builder"
            },
            agent: agent
        }, response => {
            if (response.statusCode >= 400) {
                callback(new Error(`Cannot download "${url}", status ${response.statusCode}: ${response.statusMessage}`));
                return;
            }
            const redirectUrl = response.headers.location;
            if (redirectUrl != null) {
                if (redirectCount < (_electronBuilderHttp || _load_electronBuilderHttp()).maxRedirects) {
                    this.doDownload(redirectUrl, destination, redirectCount++, options, agent, callback);
                } else {
                    callback(new Error(`Too many redirects (> ${(_electronBuilderHttp || _load_electronBuilderHttp()).maxRedirects})`));
                }
                return;
            }
            (0, (_electronBuilderHttp || _load_electronBuilderHttp()).configurePipes)(options, response, destination, callback);
        });
        this.addTimeOutHandler(request, callback);
        request.on("error", callback);
        request.end();
    }
    doApiRequest(options, token, requestProcessor) {
        let redirectCount = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : 0;

        if ((_electronBuilderHttp || _load_electronBuilderHttp()).debug.enabled) {
            (0, (_electronBuilderHttp || _load_electronBuilderHttp()).debug)(`HTTPS request: ${JSON.stringify(options, null, 2)}`);
        }
        if (token != null) {
            options.headers.authorization = token.startsWith("Basic") ? token : `token ${token}`;
        }
        return new (_bluebirdLstC2 || _load_bluebirdLstC2()).default((resolve, reject, onCancel) => {
            const request = (_https || _load_https()).request(options, response => {
                try {
                    this.handleResponse(response, options, resolve, reject, redirectCount, token, requestProcessor);
                } catch (e) {
                    reject(e);
                }
            });
            this.addTimeOutHandler(request, reject);
            request.on("error", reject);
            requestProcessor(request, reject);
            onCancel(() => request.abort());
        });
    }
}exports.NodeHttpExecutor = NodeHttpExecutor;
